### PR TITLE
Migration: Save dashboard modals

### LINKED ISSUE #22358: Migrations: Dashboard save modals

### PR DESCRIPTION
**Pull request by dprokop**
Closes https://github.com/grafana/grafana/issues/22358

So, this is a bigger endeavour than I have estimated.

Couple of important changes I have made:

1.  Added ModalsContext to be able to imperatively show modals. See https://github.com/grafana/grafana/commit/84556666324a5d686e02efa6b8de192480e432e3 for implementation and i.e. https://github.com/grafana/grafana/compare/save-dash-modals?expand=1#diff-4785772a31483292d15d9597f07853f2 for usage. 
2. Created saveDashboardButton and saveDashboardAsButton angular2react wrappers to be able to render save/save as modals in dashboard settings. They have this ugly implementation that wraps them in Provider for Redux context to be available(for navigation), but this is temporary until we migrate dashboard settings to React
3. Removed save logic from DashboardSave in favor of useDashboardSave hook ...

TODO:
- [x] Make cmd+s display Save dashboard modal. Not sure yet how to do this..
- [x] Implement Copy to clipboard and save JSON in provisioned dashboard save modal
- [x] Implement modals tests
- [x] Tweak the UI, as new form styles and big modals doesn't play well together
- [x] Fix folder picker select being cut by modal content
- [ ] Add docs for ModalsContext
- [x] Dashboard name exists modal shows but I also get an application notification error
- [x] Overwriting a dashboard with same name does not work, the overwrite confirmation modal does not go away
- [x] Align Save as... button in dashboard settings

Questions: @torkelo how important it is in the provisioned dashboard save to be able to edit dashboard JSON? In this implementation it's JSONFormatter that's used to display the dashboard model

### CODE REVIEW

File: public/app/core/components/modals/SaveDashboard/SaveDashboardErrorProxy.tsx [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+                    showModal(SaveDashboardAsModal, {
+                      dashboard,
+                      onClose: hideModal,
+                    });
+                  }}
+                >
+                  Save as
+                </Button>
+              );
+            }}
```
Comment (torkelo) [inline public/app/core/components/modals/SaveDashboard/SaveDashboardErrorProxy.tsx]: can't we re-use the SaveAsButton here?

Comment (dprokop) [reply public/app/core/components/modals/SaveDashboard/SaveDashboardErrorProxy.tsx]: We can, added the button later :)

File: public/app/core/components/modals/SaveDashboard/SaveDashboardModalProxy.tsx [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    dashboard,
+    onClose,
+    onSaveSuccess,
+  };
+  return (
+    <>
+      {isChanged && !isProvisioned && <SaveDashboardModal {...modalProps} />}
+      {isProvisioned && <SaveProvisionedDashboard {...modalProps} />}
+      {isNew && <SaveDashboardAsModal {...modalProps} isNew />}
+    </>
```
Comment (torkelo) [inline public/app/core/components/modals/SaveDashboard/SaveDashboardModalProxy.tsx]: Nice! :heart:

File: public/app/core/components/modals/SaveDashboard/SaveDashboardErrorProxy.tsx [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        <ConfirmPluginDashboardSaveModal dashboard={dashboard} onClose={onClose} />
+      )}
+    </>
+  );
+};
+
+const ConfirmPluginDashboardSaveModal: React.FC<SaveDashboardModalProps> = ({ onClose, dashboard }) => {
+  const theme = useTheme();
+  const { onDashboardSave } = useDashboardSave(dashboard);
+  const getStyles = stylesFactory((theme: GrafanaTheme) => ({
```
Comment (dprokop) [inline public/app/core/components/modals/SaveDashboard/SaveDashboardErrorProxy.tsx]: TODO: get this out!

**Review by torkelo — COMMENTED** [review]
The color of the primary button, we can't merge this in 6.7 without changing primary action to blue. Question is should we do this in 6.7 or wait with this big color change to 7.0?

image

Also the save as button is not aligned (full width) ^

image

- Folder picker has double labels
- The placement of the buttons feel off or to much bottom padding?, not sure. 
- The current modal is using a different max width, the save & save as modals are smaller. Makes them look a bit simpler and cleaner. 

Bugs:
- CMD/CTRL+S does not work anymore 
- Dashboard name exists modal shows but I also get an application notification error (these are suppressed in the current implementation). This was done by this line https://github.com/grafana/grafana/blob/master/public/app/features/dashboard/services/DashboardSrv.ts#L98  that I think the new implementation is missing
- Overwriting a dashboard with same name does not work, the overwrite confirmation modal does not go away

**Comment by dprokop** [comment]
@torkelo thanks for testing this out! Still quite some work to be done here, tracked in PR's todos

**Comment by dprokop** [comment]
@torkelo @peterholmberg @tskarhed this is now ready for the review

File: public/app/core/components/modals/SaveDashboard/forms/SaveDashboardForm.tsx [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  message: string;
+  saveVariables: boolean;
+  saveTimerange: boolean;
+}
+
+export const SaveDashboardForm: React.FC<SaveDashboardFormProps> = ({ dashboard, onCancel, onSuccess, onSubmit }) => {
+  const hasTimeChanged = useMemo(() => dashboard.hasTimeChanged(), [dashboard]);
+  const hasVariableChanged = useMemo(() => dashboard.hasVariableValuesChanged(), [dashboard]);
+
+  console.log('hasTimeChanged', hasTimeChanged);
```
Comment (torkelo) [inline public/app/core/components/modals/SaveDashboard/forms/SaveDashboardForm.tsx]: console.log :)

File: public/app/core/components/modals/SaveDashboard/forms/SaveDashboardForm.tsx [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  return (
+    <Forms.Form
+      onSubmit={async (data: SaveDashboardFormDTO) => {
+        const result = await onSubmit(dashboard.getSaveModelClone(data), data, dashboard);
+        if (result.status === 'success') {
+          console.log(data);
```
Comment (torkelo) [inline public/app/core/components/modals/SaveDashboard/forms/SaveDashboardForm.tsx]: console.log

File: public/app/core/services/keybindingSrv.ts [hunk 6]
```diff
@@ -175,6 +181,32 @@ export class KeybindingSrv {
...
+          dashboard,
+          onClose: () => {
+            ReactDOM.unmountComponentAtNode(reactRoot);
+            reactRoot.remove();
+          },
+        },
+      }),
+      reactRoot
+    );
+  };
```
Comment (torkelo) [inline public/app/core/services/keybindingSrv.ts]: is there a way we can do this with an event instead like before? so we only open the modal from one place? This feels not as great as simple sending an event, and having one owner responsible for opening and rendering the modal.

**Review by torkelo — CHANGES_REQUESTED** [review]
Great job! Would like to move all the modals and related components to features/dashboards instead (either under features/dashboard/modals or under components), as having it in core feels strange as almost everything else is under features/dashboard

File: public/app/core/services/keybindingSrv.ts (hunk 6 above)
Comment (dprokop) [reply public/app/core/services/keybindingSrv.ts]: Yes, you are right. I'm gonna do that.

**Review by torkelo — APPROVED** [review]
Looks good! great job!

**Comment by dprokop** [comment]
e2e failing tho, looking into this

File: public/app/features/dashboard/components/SaveDashboard/SaveDashboardButton.tsx [hunk 7]
```diff
@@ -0,0 +1,87 @@
+import React, { ButtonHTMLAttributes } from 'react';
```
Comment (dprokop) [inline public/app/features/dashboard/components/SaveDashboard/SaveDashboardButton.tsx]:
```suggestion
import React from 'react';
```

File: packages/grafana-ui/src/components/Forms/Button.tsx [hunk 8]
```diff
@@ -127,7 +134,7 @@ export const getButtonStyles = stylesFactory(({ theme, size, variant }: StylePro
 });
 
 // These are different from the standard Button where there are more variants.
-export type ButtonVariant = 'primary' | 'secondary' | 'destructive' | 'link';
+export type ButtonVariant = 'primary' | 'secondary' | 'destructive' | 'link' | 'primary-legacy';
```
Comment (hugohaggmark) [inline packages/grafana-ui/src/components/Forms/Button.tsx]: Does this mean are we adding `legacy` styles to new Form styles?

Comment (dprokop) [reply packages/grafana-ui/src/components/Forms/Button.tsx]: 🤦‍♂ need to remove this. Good catch

**Comment by torkelo** [comment]
Can this be merged?

**Comment by dprokop** [comment]
@torkelo just merged, sorry was off the grid last few days.

**Pull request merged**